test/Streamly/Test/Internal/Data/Unfold.hs

```diff
@@ -0,0 +1,262 @@
+-- |
+-- Module      : Main
+-- Copyright   : (c) 2019 Composewell Technologies
+--
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+module Main (main) where
+
+import qualified Streamly.Internal.Data.Unfold as UF
+import qualified Streamly.Internal.Data.Stream.IsStream as S
+import qualified Prelude as P
+
+import Control.Monad.Trans.State.Strict
+import Data.Functor.Identity
+import Prelude hiding (const, take, drop, concat)
+import Streamly.Prelude (SerialT)
+-- import System.IO.Unsafe
+import Test.Hspec as H
+import Test.Hspec.QuickCheck
+import Test.QuickCheck
+import Test.QuickCheck.Function
+
+fromStream :: Property
+fromStream =
+    property $ \x -> do
+        let strm = S.fromList x :: SerialT IO Int
+        x1 <- S.toList $ S.unfold UF.fromStream strm
+        x1 `shouldBe` (x :: [Int])
+
+nilM :: Bool
+nilM = evalState action undefined
+    where
+    action = do
+        x <- S.toList $ S.unfold (UF.nilM put) (1 :: Int)
+        y <- get
+        return $ x == ([] :: [Int]) && y == 1
+
+consM :: Bool
+consM = evalState action 0
+  where
+    action = do
+        x <-
+            S.toList $
+            S.unfold
+                (UF.consM (\a -> modify (+ a) >> get) $
+                 UF.consM (\a -> modify (+ a) >> get) $ UF.nilM put)
+                1
+        y <- get
+        return $ x == ([1, 2] :: [Int]) && y == 1
+
+-- XXX This can be removed. The types provide correctness in some sense
```

Comment:
We need tests even if it is a trivial one. One reason for that is e.g. if we accidentally remove the API from the code we won't even get to know unless we have a test.